design/recovery-internals.md

```diff
@@ -0,0 +1,193 @@
+
+
+## Phase 4: RECOVERY_TRANSACTION
+
+Not every FDB role participates in the recovery phases 1-3. This phase tells the other roles about the recovery information and triggers the recovery of those roles when necessary. 
+
+
+Storage servers (SSes) are not involved in the recovery phase 1 - 3. To notify SSes about the recovery, the master commits a recovery transaction, the first transaction in the new generation, which contains the txnStateStore information. Once storage servers receive the recovery transaction, it will compare its latest data version and the recovery version, and rollback to the recovery version if its data version is newer. Note that storage servers may have newer data than the recovery version because they pre-fetch mutations from tLogs before the mutations are durable to reduce the latency to read newly written data.
+
+
+Proxies haven’t recovered the transaction system state and cannot accept transactions yet. The master recovers proxies’ states by sending the txnStateStore to proxies through proxies’ (`txnState `) interfaces in `sendIntialCommitToResolvers()` function. Once proxies have recovered their states, they can start processing transactions. The recovery transaction that was waiting on proxies will be processed.
+
+
+The resolvers haven’t known the recovery version either. The master needs to send the lastEpochEnd version (i.e., last commit of the previous generation) to resolvers via resolvers’ (resolve) interface.
+
+
+At the end of this phase, every role should be aware of the recovery and start recovering their states.
+
+
+## Phase 5: WRITING_CSTATE
+
+Coordinators store the transaction systems’ information. The master needs to write the new tLogs into coordinators’ states to achieve consensus and fault tolerance. Only when the coordinators’ states are updated with the new transaction system’s configuration will the cluster controller tell clients about the new transaction system (such as the new proxies). 
+
+The master only needs to write the new tLogs to a quorum of coordinators for a running cluster. The only time the master has to write all coordinators is when creating a brand new database.
+
+
+Once the cstate is written, the master sets the cstateUpdated promise and moves to the ACCEPTING_COMMITS phase.
+
+
+The cstate update is done in `trackTlogRecovery()` actor.
+The actor keeps running until the recovery finishes the FULLY_RECOVERED phase. 
+The actor needs to update the cstates at the following phases:
+ALL_LOGS_RECRUITED, STORAGE_RECOVERED, and FULLY_RECOVERED. 
+For example, when the old tLogs are no longer needed, the master will write the coordinators’ state again.
+
+
+Now the main steps in recovery have finished. The master keeps waiting for all tLogs to join the system and for all storage servers to roll back their prefetched data before claiming the system is fully recovered. 
```

Comment:
```suggestion
Now the main steps in recovery have finished. The master keeps waiting for all tLogs to join the system and for all storage servers to roll back their prefetched, *uncommitted* data before claiming the system is fully recovered. 
```

Reply:
I changed *uncommitted* to "which has not been made durable on tLog,"

Reply:
I think "which has not been made durable on tLog," is not right, and "uncommitted" is the right word. The reason is that SS peeks mutations ahead and doesn't know what happened on TLog. The mutation **can** be durable on one or all of the TLogs. 

After a recovery, the master contacts the whole set of TLog servers and obtains their durable versions (actually `knownCommittedVersion`) and picks the maximum one as the recovery version. Note `knownCommittedVersion` is the version proxy told TLog servers that have committed, i.e., durable on the whole set. So when the master chooses the recovery version, all versions larger it will be discarded -- which can be the mutations SS has peeked and needs to rollback, but it's actually durable. It is just that master is not sure this version has been durable on the whole set of TLogs.